### PR TITLE
Panoptes auth

### PR DESCRIPTION
**Pull request by simoneduca**
Struggling to find the correct flow for keeping Firebase and Panoptes user in sync.
Atm the user can login/out Panoptes but it's locked out of Firebase.

### CODE REVIEW

**Comment by camallen** [comment]
@simoneduca happy to help sort this out with you.

**Comment by simoneduca** [comment]
@rogerhutchings this should be ready to go.

**Comment by simoneduca** [comment]
By the way, the hound is complaining mostly about console statements. As for the dependencies I can open a new PR later.

File: src/actions/issue.js [hunk 1]
```diff
@@ -27,7 +25,7 @@ export function fetchIssuesFromGH() {
         type: types.RECEIVE_ISSUES_ERROR_GH,
         payload: response,
       }));
```
Comment (rogerhutchings) [inline src/actions/issue.js]: You should always end a promise chain with a .`catch()`, just in case

File: src/actions/login.js [hunk 2]
```diff
@@ -1,65 +1,129 @@
...
-      })
-      .catch(error => {
-        console.error('Error logging in: ', error);
       });
-  }
+  };
+}
+
+function setFirebaseUserDisplayname(panoptesDisplayName) {
+  firebase.auth().onAuthStateChanged((user) => {
```
Comment (rogerhutchings) [inline src/actions/login.js]: I'm not sure this is right - we shouldn't create a listener for something that only happens once on login

File: src/actions/login.js [hunk 3]
```diff
@@ -1,65 +1,129 @@
...
+      });
+    }
+  };
 }
 
-export function login() {
+
+function getFirebaseToken(token) {
+//  don't hard code the endpoint
+  return fetch('http://localhost:8080/validate?token=' + token, {
```
Comment (rogerhutchings) [inline src/actions/login.js]: this is still hardcoded

File: src/actions/login.js [hunk 4]
```diff
@@ -1,65 +1,129 @@
...
-      dispatch({
-        type: types.USERVOTES_ADD,
-        payload: voteData
-      });
+      payload: user,
     });
-  }
+    if (user) {
+      userRef = firebase.database().ref(`users/${user.uid}`);
+      userListener = userRef.on('value', dataSnapshot => {
```
Comment (rogerhutchings) [inline src/actions/login.js]: This is also setting up a listener, shouldn't it be a one off?

File: src/actions/login.js (hunk 2 above)
Comment (simoneduca) [inline src/actions/login.js]: You're right I can use `getState()` instead

File: src/actions/login.js (hunk 3 above)
Comment (simoneduca) [inline src/actions/login.js]: I'd probable need a new pr for that. I don't remember where it's hosted.

File: src/actions/login.js (hunk 4 above)
Comment (simoneduca) [inline src/actions/login.js]: This should ok if I remove `onAuthStateChanged` below, right?

**Comment by rogerhutchings** [comment]
Okay, this seems to work!

- Login button doesn't change to logout
- The same user store in Redux holds the firebase user and panoptes user at different times. They should probably have their own properties inside user, it may not be clear which is being referred to at a given time in the code

**Comment by simoneduca** [comment]
Excellent, thanks. That's what I'm working on fixing atm

File: src/actions/login.js [hunk 5]
```diff
@@ -1,65 +1,129 @@
...
+        });
+      });
+    }
+  };
 }
 
-export function login() {
+
+function getFirebaseToken(token) {
+  return fetch('https://firebase-token-generator.zooniverse.org/validate', {
```
Comment (rogerhutchings) [inline src/actions/login.js]: This probably shouldn't be hardcoded either

Comment (simoneduca) [inline src/actions/login.js]: 👍

**Comment by simoneduca** [comment]
I was also thinking about how to split config values based on env var.

**Comment by rogerhutchings** [comment]
lgtm

**Pull request merged**